runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/RemoteBundle.java

```diff
@@ -40,13 +41,10 @@
   FnDataReceiver<WindowedValue<InputT>> getInputReceiver();
 
   /**
-   * {@inheritDoc}.
-   *
-   * <p>Closes this bundle. This causes the input {@link FnDataReceiver} to be closed (future calls
-   * to that {@link FnDataReceiver} will throw an exception), and causes the {@link RemoteBundle} to
-   * produce any buffered outputs. The call to {@link #close()} will block until all of the outputs
-   * produced by this bundle have been received.
+   * Closes this bundle and returns its response. This causes the input {@link FnDataReceiver} to be
+   * closed (future calls to that {@link FnDataReceiver} will throw an exception), and causes the
+   * {@link RemoteBundle} to produce any buffered outputs. The call to {@link #close()} will block
+   * until all of the outputs produced by this bundle have been received.
    */
-  @Override
-  void close() throws Exception;
+  ProcessBundleResponse close() throws Exception;
```

Comment:
Instead of providing the ProcessBundleResponse, we should expose an abstraction of a split/checkpoint, it could be part of the close() or it could be its own method like getPendingSplits and close() would throw if the caller didn't call getPendingSplits first.

Reply:
+1 to Luke's first suggestion. Requiring users to call `getPendingSplits` first seems error prone, but we should return the appropriate split/checkpoint.

In addition to this, I think it may be worth renaming `close` as it no longer follows standard `close` semantics.

Reply:
Ah, yes. That's even better and requires callers to consider split handling up-front. We can provide a default SplitHandler implementation that simply throws an unsupported exception when called.

Reply:
Added a BundleProgressHandler passed when creating the bundle, that seems a convenient place for callers to construct it.